### PR TITLE
Phenotype class attrib API -e101

### PR DESCRIPTION
**Pull request by ima23**
adding a few more methods to the api:
- AttributeAdaptor: attrib_values_for_attrib_type_code + corresponding test
- PhenotypeAdaptor: fetch_phenotype_class_types + corresponding test
- PhenotypeFeatureAdaptor: fetch_default_phenotype_classes + corresponding test

other updates:
- PhenotypeAdaptor: modify fetch to get class_attrib value in addition to class_attrib_id + corresponding test
- PhenotypeFeatureAdaptor: refactor use_phenotype_classes, existing method test pass
- Constants: add ATTRIB_TRAIT constant
- Phenotype: use ATTRIB_TRAIT constant as default class_attrib + update corresponding tests phenotype.t and phenotype_feature.t

### CODE REVIEW

**Comment by ima23** [comment]
Tests are passing: [Travis PR569](https://travis-ci.org/github/Ensembl/ensembl-variation/builds/667250805?utm_medium=notification&utm_source=github_status).

File: modules/Bio/EnsEMBL/Variation/DBSQL/AttributeAdaptor.pm [hunk 1]
```diff
@@ -83,6 +83,53 @@ sub attrib_id_for_type_value {
...
+
+    $sth->execute;
+
+    while (my ($attrib_id, $type, $value) = $sth->fetchrow_array) {
+        $attribs->{$attrib_id}->{type}  = $type;
+        $attribs->{$attrib_id}->{value} = $value;
+        $attrib_ids->{$type}->{$value} = $attrib_id;
+    }
+
+    $self->{attribs}    = $attribs;
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/AttributeAdaptor.pm]: Is this needed? It looks like you only return from $self->{attrib_ids}

File: modules/Bio/EnsEMBL/Variation/DBSQL/AttributeAdaptor.pm [hunk 2]
```diff
@@ -83,6 +83,53 @@ sub attrib_id_for_type_value {
...
+
+    my $sql = qq{
+        SELECT  a.attrib_id, t.code, a.value
+        FROM    attrib a, attrib_type t
+        WHERE   a.attrib_type_id = t.attrib_type_id
+    };
+
+    my $sth = $self->prepare($sql);
+
+    $sth->execute;
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/AttributeAdaptor.pm]: can the $attrib_type_code not be included in the execute?

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm [hunk 3]
```diff
@@ -236,6 +240,28 @@ sub _obj_from_row {
                                    mapping_type   => $row->{mapping_type} }) if defined $row->{accession} ;
 }
 
+=head2 fetch_phenotype_class_types
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm]: fetch means that you get a Phenotype object back. I would use get here

Comment (at7) [reply modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm]: get_all_phenotype_class_types

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm [hunk 4]
```diff
@@ -236,6 +240,28 @@ sub _obj_from_row {
...
+
+  Example    : $phenotype_classes = $pheno_adaptor->fetch_phenotype_class_types();
+  Description: Retrieves the list of existing Phenotype class attribs
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm]: Not sure why Phenotype is with starts with capital P

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm [hunk 5]
```diff
@@ -236,6 +240,28 @@ sub _obj_from_row {
...
+               If no phenotype class type exists undef is returned.
+  Returntype : list ref of string
+  Exceptions : none
+  Caller     : general
+=cut
+
+sub fetch_phenotype_class_types {
+  my $self = shift;
+
+  my $phenos= $self->generic_fetch();
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm]: Are you fetching all phenotypes here?

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm [hunk 6]
```diff
@@ -1706,6 +1704,29 @@ sub _get_submitter_name{
 
 }
 
+=head2 fetch_defined_phenotype_classes
+
+  Example    :
+    $pfa->fetch_defined_phenotype_classes();
+
+  Description: Fetch the phenotype classes defined in the underlying database.
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm]: mention the classes would be useful

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm [hunk 7]
```diff
@@ -1706,6 +1704,29 @@ sub _get_submitter_name{
 
 }
 
+=head2 fetch_defined_phenotype_classes
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm]: Again fetch should be replaced by get_all_*

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm [hunk 8]
```diff
@@ -1706,6 +1704,29 @@ sub _get_submitter_name{
...
+
+  Example    :
+    $pfa->fetch_defined_phenotype_classes();
+
+  Description: Fetch the phenotype classes defined in the underlying database.
+         The default behaviour is to return phenotype features of any phenotype class.
+  Returntype : array of string
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm]: is it an array or list ref?

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm [hunk 9]
```diff
@@ -1706,6 +1704,29 @@ sub _get_submitter_name{
...
+  Exceptions : none
+  Caller     : general
+  Status     : Stable
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm]: inconsistent use of capitalisation

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm [hunk 10]
```diff
@@ -1706,6 +1704,29 @@ sub _get_submitter_name{
...
+
+=cut
+
+sub fetch_defined_phenotype_classes {
+  my $self = shift;
+
+  if (!defined($self->{phenotype_classes})) {
+    $self->{phenotype_classes} = $self->db->get_AttributeAdaptor->attrib_values_for_attrib_type_code(ATTRIB_TYPE_PHENOTYPE_TYPE);
```
Comment (at7) [inline modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeFeatureAdaptor.pm]: I would check that the adaptor exists and is available here

File: modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm (hunk 5 above)
Comment (ima23) [reply modules/Bio/EnsEMBL/Variation/DBSQL/PhenotypeAdaptor.pm:256]: Yes, currently I am. I was debating it I should do a distinct sql select, can change it to a sql.

**Comment by at7** [comment]
merged into master and release/101

**Pull request merged**